File: packages/astro-engine-python/src/astro_engine/location_distance.py
"""location.distance — great-circle distance for arbitrary coordinates."""

from __future__ import annotations

import math
from typing import Any, Mapping

from astro_engine.errors import ValidationError
from astro_engine.grid import EARTH_RADIUS_METERS, METERS_PER_MILE
from astro_engine.validate import require_finite_number

CAPABILITY_ID = "location.distance"


def distance_miles(from_latitude: float, from_longitude: float,
                   to_latitude: float, to_longitude: float) -> float:
    """Shortest spherical arc, using the same Earth radius as location.grid."""
    lat1, lat2 = math.radians(from_latitude), math.radians(to_latitude)
    dlat = lat2 - lat1
    dlon = math.radians(to_longitude - from_longitude)
    haversine = (math.sin(dlat / 2.0) ** 2
                 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2.0) ** 2)
    haversine = min(1.0, max(0.0, haversine))
    angle = 2.0 * math.atan2(math.sqrt(haversine), math.sqrt(1.0 - haversine))
    return angle * EARTH_RADIUS_METERS / METERS_PER_MILE


def _coordinate(value: Any) -> tuple[float, float]:
    if not isinstance(value, Mapping) or set(value) != {"latitude", "longitude"}:
        raise ValidationError("invalid location.distance input")
    try:
        latitude = require_finite_number(value["latitude"], "latitude")
        longitude = require_finite_number(value["longitude"], "longitude")
    except ValidationError as exc:
        raise ValidationError("invalid location.distance input") from exc
    if not -90 <= latitude <= 90 or not -180 <= longitude <= 180:
        raise ValidationError("invalid location.distance input")
    return latitude, longitude


def location_distance(input: Mapping[str, Any] | Any) -> dict[str, float]:
    if not isinstance(input, Mapping) or set(input) != {"from", "to"}:
        raise ValidationError("invalid location.distance input")
    start = _coordinate(input["from"])
    end = _coordinate(input["to"])
    return {"distance_miles": distance_miles(*start, *end)}

File: packages/astro-engine-swift/Sources/AstroEngine/Location/LocationDistance.swift
import Foundation
import CoreFoundation

public enum LocationDistance {
    public static let capabilityID = "location.distance"

    /// Shortest spherical arc in miles, using location.grid's Earth radius.
    public static func miles(from start: Coordinate, to end: Coordinate) -> Double {
        let lat1 = start.latitude * .pi / 180
        let lat2 = end.latitude * .pi / 180
        let deltaLat = lat2 - lat1
        let deltaLon = (end.longitude - start.longitude) * .pi / 180
        let raw = pow(sin(deltaLat / 2), 2)
            + cos(lat1) * cos(lat2) * pow(sin(deltaLon / 2), 2)
        let haversine = min(1, max(0, raw))
        let angle = 2 * atan2(sqrt(haversine), sqrt(1 - haversine))
        return angle * GeographicGridGenerator.earthRadiusMeters
            / GeographicGridGenerator.metersPerMile
    }

    public static func evaluate(_ input: [String: Any]) throws -> [String: Any] {
        guard Set(input.keys) == ["from", "to"] else {
            throw LocationDistanceError.invalidInput
        }
        let start = try coordinate(input["from"])
        let end = try coordinate(input["to"])
        return ["distance_miles": miles(from: start, to: end)]
    }

    private static func coordinate(_ value: Any?) throws -> Coordinate {
        guard let row = value as? [String: Any],
              Set(row.keys) == ["latitude", "longitude"] else {
            throw LocationDistanceError.invalidInput
        }
        let latitude = try number(row["latitude"])
        let longitude = try number(row["longitude"])
        guard (-90...90).contains(latitude), (-180...180).contains(longitude) else {
            throw LocationDistanceError.invalidInput
        }
        return Coordinate(latitude: latitude, longitude: longitude)
    }

    private static func number(_ value: Any?) throws -> Double {
        guard let number = value as? NSNumber,
              CFGetTypeID(number) != CFBooleanGetTypeID(),
              number.doubleValue.isFinite else {
            throw LocationDistanceError.invalidInput
        }
        return number.doubleValue
    }
}

public enum LocationDistanceError: Error {
    case invalidInput
}
